### PR TITLE
Refactor flatbuffer_direct request, lowering, and pass-state boundaries

### PR DESCRIPTION
**Pull request by PINTO0309**
## Summary

This pull request continues the staged refactoring of the TensorFlow-free `flatbuffer_direct` pipeline. It makes three related areas smaller and safer without changing the public CLI/Python API or the existing conversion order:

1. generated PyTorch-source repair policy is extracted from the large exporter orchestrator into focused, Torch-free policy functions;
2. normalized conversion requests, requested-artifact controls, Session-owned graph state, and lowering mutation APIs become explicit internal boundaries;
3. adjacent registered ModelIR passes reuse one lazily constructed differential graph index when no untracked mutation occurs between them.

The branch contains 13 coherent commits and changes 22 files (`+2,062 / -555`). The work is intentionally incremental: existing rule order, match grammar, defaults, output names, return structures, diagnostics, exception behavior, and artifact semantics are preserved. No dependency was added, and normal direct TFLite conversion remains free of TensorFlow imports.

## Why this change is needed

The direct conversion pipeline has accumulated many interacting rules, repeated option parsing, and repeated graph-state construction. That creates three practical risks:

- a local rule change can accidentally alter later rule eligibility because ordering and state mutation are implicit;
- optional exporters can influence unrelated conversions by parsing settings even when their artifacts were not requested;
- graph producer/consumer/layout state can become stale or be rebuilt repeatedly during lowering and post-lowering passes.

This pull request reduces those risks through narrow ownership boundaries and characterization tests. It does not attempt a single large rewrite of the full converter.

### 1. Extracted ordered PyTorch fast-precanonicalization policy

Six behavior-sensitive rule families were moved from `pytorch_exporter.py` into `pytorch_fast_precanonicalize_policy.py`:

- downstream-evidence repair for aligned binary shapes;
- Resize repair using input layout and direct or reshaped BatchNorm evidence;
- aligned BatchNorm constant normalization;
- Local Response Normalization layout and static-shape propagation;
- common recording of literal rewritten static shapes;
- NHWC AveragePool bridge state updates.

The exporter still owns orchestration and preserves the original scan order. The new policy owner contains the match, guard, rewrite, and state-update decisions for each family. This keeps the remaining exporter code readable while preventing policy logic and generated-source parsers from being duplicated.

Important preserved behaviors include:

- general binary and Resize repair still run before their narrower evidence-based fallbacks;
- former loop `continue` behavior is represented by an explicit short-circuit result;
- only the same narrow generated-statement grammar is accepted;
- direct BatchNorm constants still require registered matching channel evidence;
- already-reshaped BatchNorm constants retain their distinct, narrower legacy guard;
- LRN propagation remains state-only and does not mark source text as changed;
- dynamic or unparseable shapes do not replace known static-shape state;
- NHWC AveragePool bridge layout and shape updates remain at the same point in the ordered scan.

The audited fast-precanonicalization orchestrator is now 294 lines, down from 482 lines at the start of this continuation.

### 2. Enforced one normalized request boundary

`ConversionRequest.from_kwargs` is now the only raw-`kwargs` boundary in the direct exporter. All downstream reads use the immutable normalized request mapping or the typed `ArtifactPlan`.

This mechanically replaced 36 raw option reads while retaining:

- all existing option names and default values;
- current coercion and validation behavior;
- current downstream arguments and return values;
- existing public API and exception behavior.

An AST-based architecture test prevents raw `kwargs` access from being reintroduced below this boundary.

### 3. Avoided preparation of unrequested artifacts

`artifact_preparation.py` now resolves exporter-specific controls only for requested artifacts. It owns the settings for:

- SavedModel and PyTorch output paths;
- PyTorch model persistence;
- native-code-generation timeout;
- input-shape and test-data settings;
- custom calibration input data;
- quantization type;
- input and output quantization dtypes.

A TFLite-only request therefore does not parse unused SavedModel, PyTorch, or calibration settings. For example, an invalid unused PyTorch timeout can no longer fail an unrelated TFLite-only conversion. Requested artifacts keep the existing paths, defaults, optional dependencies, and behavior. Resolved control mappings are immutable.

### 4. Restored Session-owned consumer counts

The `ConversionSession` migration had eliminated a duplicate ONNX scan, but `LoweringContext` was still receiving an empty compatibility consumer-count dictionary. The lowering boundary now receives `ConversionSession.tensor_consumer_count`, backed by the authoritative `GraphIndex`.

This restores the fan-out safety contract used by inverse-Transpose elision, including repeated occurrences of the same input tensor, without rebuilding another whole-graph consumer map.

### 5. Synchronized lowering-time layout mutations

`LoweringContext.set_tensor_layout()` is now the single lowering-time mutation boundary for logical and physical layout. It updates both `TensorIR` metadata and the Session-owned `LayoutState` immediately.

Shape-family edge-Pad passthroughs, integer-linear Resize casts, and rank-three Resize adapters no longer assign layout fields directly. This fixes observed stale layout state before the first post-lowering pass without adding an eager full-ModelIR synchronization scan.

An architecture test rejects future direct logical/physical layout assignment from op builders.

### 6. Protected synthetic inverse-Transpose fan-out

`LoweringContext.add_operator()` and `remove_operator()` now maintain differential producer and consumer counts for the partial ModelIR. Operator removal is centralized at this boundary.

Inverse-Transpose elision now distinguishes:

- an ONNX edge, whose authoritative count comes from the Session graph index; and
- a synthetic edge, whose count comes from the current partial ModelIR plus the pending inverse use.

An exclusive inverse pair is still removed. If an already-emitted synthetic side consumer needs the producer, the producer is retained. This fixes stale producer-index and dead-edge risks without scanning the partial graph after every emitted operator.

An architecture test prevents op builders from directly mutating `model_ir.operators`.

### 7. Reused pass state across an adjacent registered-pass cluster

A lazy, identity-bound `ModelIRPassStateScope` now allows adjacent registered pass groups to share one differentially maintained `ModelIRGraphIndex` and `LayoutState`.

The first production use covers six repeated lowering clusters while preserving the exact existing runner order:

1. transpose/Mean cleanup;
2. Mean/Mul/Add/Conv attention cleanup;
3. optional LayerNorm cleanup;
4. terminal Mean cleanup;
5. squeeze-and-excitation convolution cleanup;
6. squeeze-and-excitation fully connected cleanup;
7. optional convolution-attention cleanup.

The scope is deliberately bounded: it ends before every legacy helper that mutates ModelIR outside the differential index. It is also lazy, so if every model-only preflight misses, no graph index is constructed. Depending on enabled optional passes, a cluster that previously constructed up to seven identical indexes now constructs at most one.

Diagnostics retain their existing meaning: only the group that actually constructs state reports `state_built: true`; later groups reusing it report `false`, so aggregate build counts remain accurate.

## Compatibility and dependency boundaries

This pull request intentionally preserves:

- existing CLI and Python API signatures;
- default backend and option defaults;
- artifact names, paths, return structures, and report formats;
- generated-source rule order and matching eligibility;
- quantization defaults and requested-artifact semantics;
- optional TensorFlow and PyTorch behavior when explicitly requested;
- TensorFlow-free normal `flatbuffer_direct` TFLite conversion.

No package, optional extra, or generated schema was added.

## Validation

All recorded validation used `uv`, a single pytest process, and single-process inference where artifact execution was involved. Host `PYTHONPATH` and `LD_LIBRARY_PATH` were removed for the focused environment where required.

### Policy extraction and architecture ownership

```text
env -u PYTHONPATH -u LD_LIBRARY_PATH \
  OMP_NUM_THREADS=1 MKL_NUM_THREADS=1 \
  uv run pytest -q \
  tests/test_flatbuffer_direct_pytorch_fast_precanonicalize_policy.py \
  tests/test_flatbuffer_direct_architecture.py

124 passed
```

This selection covers positive rewrites and no-op guards for downstream evidence, direct and reshaped BatchNorm constants, Resize layout evidence, LRN propagation, literal static-shape recording, NHWC Pool bridges, channel mismatches, mixed layouts, dynamic shapes, and short-circuit behavior.

### Normalized request boundary

```text
uv run pytest -q \
  tests/test_flatbuffer_direct_core.py \
  tests/test_flatbuffer_direct_architecture.py

126 passed
```

The AST gate proves that the exporter has zero `kwargs.get` calls and exactly one raw `kwargs` read: the argument to `ConversionRequest.from_kwargs`.

### Requested-artifact preparation

```text
uv run pytest -q \
  tests/test_flatbuffer_direct_artifact_preparation.py \
  tests/test_flatbuffer_direct_core.py \
  tests/test_flatbuffer_direct_architecture.py

145 passed
```

A raising mapping proves that unrequested exporter and quantization settings are not accessed. Requested values, legacy defaults, calibration-only inputs, timeout coercion, and immutable mappings are covered.

### Session consumer-count handoff

```text
uv run pytest -q \
  tests/test_flatbuffer_direct_core.py \
  tests/test_tflite_builder_direct.py::test_flatbuffer_direct_transpose_quantize_transpose_optimization \
  tests/test_tflite_builder_direct.py::test_flatbuffer_direct_transpose_quantize_transpose_fanout_optimization \
  tests/test_tflite_builder_direct.py::test_flatbuffer_direct_transpose_quantize_transpose_preserves_dynamic_batch_signature

32 passed
```

The Session fixture verifies exact consumer counts for a tensor reused by Add and Identity.

### Lowering-time layout synchronization

```text
uv run pytest -q \
  tests/test_flatbuffer_direct_core.py \
  tests/test_pad_edge_lowering.py \
  tests/test_flatbuffer_direct_resize_integer_linear.py \
  tests/test_flatbuffer_direct_architecture.py::test_op_builders_mutate_layout_only_through_lowering_context

33 passed
```

The rank-three Resize hook inspects `LayoutState` before the first post-lowering pass. Focused edge-Pad and integer-linear Resize tests also serialize and execute their TFLite artifacts sequentially.

### Synthetic fan-out protection

```text
uv run pytest -q \
  tests/test_flatbuffer_direct_core.py \
  tests/test_tflite_builder_direct.py::test_flatbuffer_direct_elides_inverse_transpose_chain_at_generation \
  tests/test_tflite_builder_direct.py::test_flatbuffer_direct_no_dead_operator_outputs_after_prune \
  tests/test_flatbuffer_direct_architecture.py::test_op_builders_mutate_operator_list_only_through_lowering_context

35 passed
```

Focused context cases prove both sides of the contract: an exclusive inverse pair is removed with its differential index entries, while a synthetic bridge consumed by another operator retains its producer.

### Adjacent pass-state reuse

```text
uv run pytest -q \
  tests/test_flatbuffer_direct_mean_layout.py \
  tests/test_flatbuffer_direct_layernorm_layout.py \
  tests/test_flatbuffer_direct_terminal_mean_layout.py \
  tests/test_flatbuffer_direct_se_layout.py \
  tests/test_flatbuffer_direct_core.py \
  tests/test_flatbuffer_direct_pass_efficiency.py \
  tests/test_flatbuffer_direct_architecture.py::test_lowerer_mean_attention_cluster_reuses_one_pass_state_scope

65 passed
```

The reuse characterization observes one `ModelIRGraphIndex.refresh()` across two matching adjacent runners and diagnostic construction flags `[true, false]`. The all-preflight-miss case observes zero refreshes and `[false, false]`. The architecture gate fixes runner order, the shared scope keyword, and all six production invocations.

Changed modules pass Python syntax compilation. Scoped Ruff checks pass for the modified owners and tests, with documented ignores only for pre-existing findings in legacy files. `git diff --check` passes.

## Known limitations and remaining work

- No Tier corpus or broad real-model conversion run was performed for these mechanical checkpoints, following the current instruction to keep conversion tests minimal. Broad model-level regression safety is therefore not claimed here.
- The optional TensorFlow suite was not synchronized or run.
- Whole-file Ruff on `pytorch_exporter.py` still reports pre-existing compatibility re-export, unused scaffold, and undefined-name findings. The changed policy owners and focused tests pass their scoped checks, and the exporter passes syntax compilation.
- The larger Goal remains incomplete: additional contiguous registered-pass clusters can reuse state only after proving that their scope contains no raw ModelIR mutation; artifact-matrix coverage and broader Tier performance/accuracy gates remain future work.

## Suggested review order

1. Review `pytorch_fast_precanonicalize_policy.py` together with its focused characterization tests, then confirm the unchanged call order in `pytorch_exporter.py`.
2. Review `ConversionRequest` usage and `artifact_preparation.py` to verify the requested-only option boundary.
3. Review the `ConversionSession` consumer-count handoff and `LoweringContext` layout/operator mutation APIs.
4. Review the inverse-Transpose synthetic fan-out regression tests.
5. Review `ModelIRPassStateScope`, the bounded Mean/attention cluster, and the index-construction characterization.
6. Review `docs/flatbuffer_direct_architecture.md` and the continuation handoff for preserved contracts and declared limitations.

## Commits

- `3ac19b40` Extract downstream binary layout repair
- `008e4ad0` Extract Resize BN evidence repair
- `80d1d6a5` Extract aligned BatchNorm constant repair
- `91a0a52d` Extract local response norm layout propagation
- `b00774a7` Centralize rewritten static shape recording
- `95fbd0cb` Co-locate NHWC bridge state updates
- `907c91fa` Enforce direct conversion request boundary
- `5848cc28` Guard optional exporter settings by request
- `e3c03e3d` Guard quantization type options by request
- `4f3d20b0` Restore session consumer counts in lowering
- `6e8a8486` Synchronize lowering layout state mutations
- `e7c1457d` Protect synthetic transpose fanout during lowering
- `2a2968cc` Reuse pass state across mean attention cluster

### CODE REVIEW

**Comment by PINTO0309** [comment]
Closing this pull request as requested. The fb-refactor5 branch will remain available, and subsequent work will be delivered through commits and pushes only; no replacement pull request will be opened.
